src/modules-load/modules-load.c

```diff
@@ -163,34 +179,42 @@ static int run(int argc, char *argv[]) {
         if (r < 0)
                 log_warning_errno(r, "Failed to parse kernel command line, ignoring: %m");
 
-        r = module_setup_context(&ctx);
-        if (r < 0)
-                return log_error_errno(r, "Failed to initialize libkmod context: %m");
+        _cleanup_close_pair_ int seq[2] = EBADF_PAIR;
+        const int max_tasks = MAX(MIN(sysconf(_SC_NPROCESSORS_ONLN) / 2, 10), 2);
+        _cleanup_free_ pthread_t *threads = malloc(sizeof(pthread_t) * max_tasks);
+        if (socketpair(AF_UNIX, SOCK_SEQPACKET, 0, seq) == -1)
+                log_error_errno(errno, "socketpair");
 
-        r = 0;
+        // Create worker threads
+        for (int i = 0; i < max_tasks; ++i)
+                if (pthread_create(&threads[i], NULL, worker_thread, &seq[1]) != 0)
+                        log_error_errno(errno, "pthread_create");
 
+        r = 0;
         if (argc > optind) {
                 for (int i = optind; i < argc; i++)
-                        RET_GATHER(r, apply_file(ctx, argv[i], false));
-
+                        RET_GATHER(r, apply_file(seq, argv[i], false));
         } else {
                 _cleanup_strv_free_ char **files = NULL;
-
-                STRV_FOREACH(i, arg_proc_cmdline_modules) {
-                        k = module_load_and_warn(ctx, *i, true);
-                        if (k == -ENOENT)
-                                continue;
-                        RET_GATHER(r, k);
-                }
+                STRV_FOREACH(i, arg_proc_cmdline_modules)
+                        if (send(seq[0], *i, strlen(*i), 0))
+                                log_error_errno(errno, "send");
 
                 k = conf_files_list_nulstr(&files, ".conf", NULL, 0, conf_file_dirs);
                 if (k < 0)
                         return log_error_errno(k, "Failed to enumerate modules-load.d files: %m");
 
                 STRV_FOREACH(fn, files)
-                        RET_GATHER(r, apply_file(ctx, *fn, true));
+                        RET_GATHER(r, apply_file(seq, *fn, true));
         }
 
+        // Close the sending socket
+        safe_close(seq[0]);
+
+        // Wait for all threads to finish
```

Comment:
here and everywhere, we use `//` style comments only for FIXME-style stuff. For explanatory comments we use `/* … */` style comments.

Also see: https://systemd.io/CODING_STYLE